File: scripts/plot_poses.m
%% housekeeping
clear; close all; clc;

%% init

pose_type = 'NLS'; % pose_type

truePosesMat = f_read_poses("../data/true_poses.csv");

solvedPosesMat = f_read_poses("../data/solved_poses.csv");

filteredPosesMat = f_read_poses("../data/filtered_poses.csv");

%statesMat = f_read_covars("../data/kf_states.csv");

%covarsMat = f_read_covars("../data/kf_covars.csv");

[num_poses,~] = size(truePosesMat);
dt = 0.1;
tVec = 0:dt:(num_poses-1)*dt;

%% compute position score

posScoreVec         = 100*ones(num_poses,1);
posScoreVecFiltered = 100*ones(num_poses,1);

for idx = 1:num_poses,

    posErr         = truePosesMat(idx,1:3) - solvedPosesMat(idx,1:3);
    posErrFiltered = truePosesMat(idx,1:3) - filteredPosesMat(idx,1:3);

    posScoreVec(idx)         = norm(posErr);%/norm(truePosesMat(idx,1:3));
    posScoreVecFiltered(idx) = norm(posErrFiltered);%/norm(truePosesMat(idx,1:3));
end

%% compute attitude score

attScoreVec         = 1000*ones(num_poses,1);
attScoreVecFiltered = 1000*ones(num_poses,1);

for idx = 1:num_poses,

    quat            = angle2quat(     truePosesMat(idx,4),     truePosesMat(idx,5),     truePosesMat(idx,6) );
    quatHat         = angle2quat(   solvedPosesMat(idx,4),   solvedPosesMat(idx,5),   solvedPosesMat(idx,6) );
    quatHatFiltered = angle2quat( filteredPosesMat(idx,4), filteredPosesMat(idx,5), filteredPosesMat(idx,6) );
    
    dquat         = quatmultiply( quatnormalize(quat), quatconj(quatnormalize(quatHat)) );
    dquatFiltered = quatmultiply( quatnormalize(quat), quatconj(quatnormalize(quatHatFiltered)) );
    %dqVec = (quat.normalized())*(quatHat.normalized().conjugate());

    attScoreVec(idx)         = 2*acos( abs( dquat(1) ) ); % rad
    attScoreVecFiltered(idx) = 2*acos( abs( dquatFiltered(1) ) ); % rad
end


%% plotting position
f1 = figure(1);
subplot(4,1,1)
plot(tVec, truePosesMat(:,1))
title('Position')
hold on
plot(tVec, solvedPosesMat(:,1))
plot(tVec, filteredPosesMat(:,1),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
legend('true',pose_type,'MEKF','Location','Northwest')
xlabel('time [s]')
ylabel('x [m]')

subplot(4,1,2)
plot(tVec, truePosesMat(:,2))
hold on
plot(tVec, solvedPosesMat(:,2))
plot(tVec, filteredPosesMat(:,2),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
%legend('true',pose_type,'MEKF','Location','Southwest')
xlabel('time [s]')
ylabel('y [m]')

subplot(4,1,3)
plot(tVec, truePosesMat(:,3))
hold on
plot(tVec, solvedPosesMat(:,3))
plot(tVec, filteredPosesMat(:,3),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
%legend('true',pose_type,'MEKF','Location','Northwest')
xlabel('time [s]')
ylabel('z [m]')
%ylim([0,50])

subplot(4,1,4)
plot(tVec, posScoreVec,'color',[0.8500, 0.3250, 0.0980])
hold on
plot(tVec, posScoreVecFiltered,'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
legend(pose_type,'MEKF')
xlabel('time [s]')
ylabel('pos\_score [m]')
ylim([0 30])

figpos = [0 0 1920/2 1080];
set(f1,'Position',figpos)
boldify;
set(gcf, 'PaperPositionMode', 'auto');

%% plotting attitude
f2 = figure(2);
subplot(4,1,1)
plot(tVec, rad2deg(truePosesMat(:,4)))
title('Attitude')
hold on
plot(tVec, rad2deg(solvedPosesMat(:,4)))
plot(tVec, rad2deg(filteredPosesMat(:,4)),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
legend('true',pose_type,'MEKF','Location','Southeast')
xlabel('time [s]')
ylabel('\phi [deg]')

subplot(4,1,2)
plot(tVec, rad2deg(truePosesMat(:,5)))
hold on
plot(tVec, rad2deg(solvedPosesMat(:,5)))
plot(tVec, rad2deg(filteredPosesMat(:,5)),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
%legend('true',pose_type,'MEKF','Location','Southeast')
xlabel('time [s]')
ylabel('\theta [deg]')

subplot(4,1,3)
plot(tVec, rad2deg(truePosesMat(:,6)))
hold on
plot(tVec, rad2deg(solvedPosesMat(:,6)))
plot(tVec, rad2deg(filteredPosesMat(:,6)),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
%legend('true',pose_type,'MEKF','Location','Southeast')
xlabel('time [s]')
ylabel('\psi [deg]')

subplot(4,1,4)
plot(tVec, rad2deg(attScoreVec),'color',[0.8500, 0.3250, 0.0980])
hold on
plot(tVec, rad2deg(attScoreVecFiltered),'color',[0.4940, 0.1840, 0.5560],'LineWidth',2)
grid on
legend(pose_type,'MEKF')
xlabel('time [s]')
ylabel('att\_score [deg]')
ylim([0 30])

figpos = [0 0 1920/2 1080];
set(f2,'Position',figpos)
boldify;
set(gcf, 'PaperPositionMode', 'auto');

File: scripts/plot_poses.jl
using DelimitedFiles
using ReferenceFrameRotations
using Plots

## init
truePosesMat = readdlm("data/true_poses.csv", ',', Float64, '\n');

solvedPosesMat = readdlm("data/solved_poses.csv", ',', Float64, '\n');

filteredPosesMat = readdlm("data/filtered_poses.csv", ',', Float64, '\n');

statesMat = readdlm("data/kf_states.csv", ',', Float64, '\n');

covarsMat = readdlm("data/kf_covars.csv", ',', Float64, '\n');

num_poses = size(truePosesMat)[1];
dt = 0.5;
tVec = 0:dt:(num_poses - 1) * dt;

## compute position score
posScoreVec         = 1000 * ones(num_poses,1);
posScoreVecFiltered = 1000 * ones(num_poses,1);

for idx = 1:num_poses,
    posErr         = truePosesMat[idx,1:3] - solvedPosesMat[idx,1:3];
    posErrFiltered = truePosesMat[idx,1:3] - filteredPosesMat[idx,1:3];

    posScoreVec[idx]         = norm(posErr);#/norm(truePosesMat[idx,1:3]);
    posScoreVecFiltered[idx] = norm(posErrFiltered);#/norm(truePosesMat[idx,1:3]);
end

## compute attitude score
attScoreVec         = 1000 * ones(num_poses, 1);
attScoreVecFiltered = 1000 * ones(num_poses, 1);

for idx = 1:num_poses
    # TODO migrate to julia quat object
    #q = angle_to_quat(pi/2, pi/4, pi/3, :ZYX)
    quat            = angle_to_quat(truePosesMat[idx,4],     truePosesMat[idx,5],     truePosesMat[idx,6], :ZYX);
    quatHat         = angle_to_quat(solvedPosesMat[idx,4],   solvedPosesMat[idx,5],   solvedPosesMat[idx,6], :ZYX);
    quatHatFiltered = angle_to_quat(filteredPosesMat[idx,4], filteredPosesMat[idx,5], filteredPosesMat[idx,6], :ZYX);

    dquat         = quat * conj(quatHat);
    dquatFiltered = quat * conj(quatHatFiltered);
    #dquat         = quatmultiply( quatnormalize(quat), quatconj(quatnormalize(quatHat)) );
    #dquatFiltered = quatmultiply( quatnormalize(quat), quatconj(quatnormalize(quatHatFiltered)) );

    attScoreVec[idx]         = 2 * acos(abs(dquat.q0)); # rad
    attScoreVecFiltered[idx] = 2 * acos(abs(dquatFiltered.q0)); # rad
end

## plotting
gr() # Choose a backend

# shared attributes
xlims_ = (0,maximum(tVec)+dt);
labels = ["true" "NLS" "KF"];

# plotting position
datap1 = [truePosesMat[:,1] solvedPosesMat[:,1] filteredPosesMat[:,1]];
posp1 = plot(tVec, datap1, label = labels, legend = :bottomleft);
xlims!(xlims_)
ylabel!("x [m]")

datap2 = [truePosesMat[:,2] solvedPosesMat[:,2] filteredPosesMat[:,2]];
posp2 = plot(tVec, datap2, label = labels, legend = :bottomleft);
xlims!(xlims_)
ylabel!("y [m]")

datap3 = [truePosesMat[:,3] solvedPosesMat[:,3] filteredPosesMat[:,3]];
posp3 = plot(tVec, datap3, label = labels, legend = :bottomleft);
xlims!(xlims_)
ylabel!("z [m]")

datap4 = [posScoreVec posScoreVecFiltered];
posp4 = plot(tVec, datap4, label = "")
xlims!(xlims_)
xlabel!("time [s]")
ylabel!("pos_score [ ]")

plot(posp1, posp2, posp3, posp4, layout = (4,1))


#=
# plotting attitude
labels = ["true" "NLS" "KF"];
datap5 = (180.0/π)*[truePosesMat[:,4] solvedPosesMat[:,4] filteredPosesMat[:,4]];
posp5 = plot(tVec, datap5,  xlims = xlims_,
                            ylabel = "ϕ [deg]",
                            label = labels);

datap6 = (180.0/π)*[truePosesMat[:,5] solvedPosesMat[:,5] filteredPosesMat[:,5]];
posp6 = plot(tVec, datap6,  xlims = xlims_,
                            ylabel = "θ [deg]",
                            label = "");

datap7 = (180.0/π)*[truePosesMat[:,6] solvedPosesMat[:,6] filteredPosesMat[:,6]];
posp7 = plot(tVec, datap7,  xlims = xlims_,
                            ylabel = "ψ [deg]",
                            label = "");

datap8 = (180.0/π)*[attScoreVec attScoreVecFiltered];
posp8 = plot(tVec, datap8,  xlims = xlims_,
                            xlabel = "time [s]",
                            ylabel = "att_score [ ]",
                            label = "");

plot(posp5, posp6, posp7, posp8, layout = (4,1))
=#
